storage/innobase/mtr/mtr0mtr.cc

```diff
@@ -1217,19 +1217,35 @@ inline void log_t::resize_write(lsn_t lsn, const byte *end, size_t len,
     if (!resize_flush_buf)
     {
       ut_ad(is_pmem());
+      lsn_lock.wr_lock();
       const size_t resize_capacity{resize_target - START_OFFSET};
-      const lsn_t resizing{resize_in_progress()};
-      if (UNIV_UNLIKELY(lsn < resizing))
       {
-        size_t l= resizing - lsn;
-        if (l >= len)
-          return;
-        end+= l - len;
-        len-= l;
-        lsn+= l;
+        const lsn_t resizing{resize_in_progress()};
+        if (UNIV_UNLIKELY(lsn < resizing))
+        {
+          size_t l= resizing - lsn;
+          if (l >= len)
+            goto pmem_done;
+          end+= l - len;
+          len-= l;
+          lsn+= l;
```

Comment:
`end+= l - len;` looks incorrect to me. When resize_lsn is ahead, we need to skip copying the initial part of the buffer i.e. `l` (resizing - lsn). Why should we subtract `len` here ?

Can we have a test covering this part ? If resize LSN is set when LSN is in the beginning of a write block (`write_size`), the subsequent mtr should follow this path. Once the resize is complete a crash and restart should reflect the issue if I am correct.

I am assuming it may not be very difficult to have a test now we are setting the resize_lsn ahead of log_sys LSN in the other patch. Perhaps this code was unreachable be before the other change.

Reply:
"The other patch" would be #3458, which only affects the non-PMEM code path. For *memory-mapped log* (which is the domain of this fix: `is_pmem()` holds if and only if `!flush_buf` and `!resize_flush_buf`), the only way how `log_sys.resize_lsn` can be less than `log_sys.lsn` is when it is 0 (online log resizing is not in progress) or 1 (`log_t::resize_start()` is creating and allocating an `ib_logfile101` that aims to be the resized log file).

It would seem that this is in fact dead code and needs to be replaced with a debug assertion.

The critical path of `log_t::resize_start()` is enclosed between `log_resize_acquire()` and `log_resize_release()`, which includes an exclusive `log_sys.latch`. The callers of this function `log_t::resize_write()` must hold a shared or exclusive `log_sys.latch` until they have completed their write. Therefore, `resize_write()` and `resize_start()` cannot run concurrently and there is no chance that a resizing that is about to start would cause `lsn < resizing` to hold or otherwise be able to conflict with an older `log_t::resize_write()`.

Reply:
Ok, let's remove the dead code then.

Reply:
Sorry, I realized that it is not dead code at all. Regular log writers would only hold a shared `log_sys.latch`, and therefore a concurrent `log_t::resize_write()` in another thread could have updated the `log_sys.resize_lsn`. If a concurrent `resize_write()` with a larger LSN was able to grab the `log_sys.lsn_lock` first, then any older writes would have to be discarded.

Reply:
The code after `goto pmem_done` is actually dead, because the `resize_lsn.store(lsn)` in another thread must have advanced the log beyond our `lsn+len`. The concurrent mini-transaction writes to `log_sys.buf` cannot overlap, so they must not overlap here for `log_sys.resize_buf` either.

Reply:
Right, Thanks for explaining.